Change ConfFile API

otaviobp opened the pull request:
Followed suggestions on https://github.com/01org/camera-streaming-daemon/pull/29

lucasdemarchi commented:
First 4 patches applied

edersondisouza commented on conf_file.cpp:
```diff
@@ -16,268 +16,416 @@
+        return s;
+
+    log_debug("ConfFile: Adding section '%.*s'", (int)(len - 2), addr + 1);
+    _sections = new section{addr + 1, len - 2, nullptr, _sections};
```
Nice thing this `_list = new item{..., _list}`

edersondisouza commented on conf_file.cpp:
```diff
@@ -16,268 +16,416 @@
+        if (s->len - 1 > section_name_len && strncaseeq(section_name, s->name, section_name_len)
+            && s->name[section_name_len] == separator) {
+            *sub_name = s->name + section_name_len + 1;
+            *sub_len = s->len - (*sub_name - s->name);
```
This ends up allowing 'unnamed' sub sections - is this desired?

otaviobp replied:
We don't allow unnamed sub sections because the if is checking if we have 2 chars (separator + 1) after the section name.
But, it is not trimming the subsection and section names, so we may have a subsection called " ".
So, do we need to trim section names, like:
[     section_name        sub_section_name       ]

otaviobp replied:
github removed my trailing spaces. heheh.  Now with spaces:

    [      section_name             sub_section_name           ]

edersondisouza replied:
Well.... we already trim everywhere =D

edersondisouza commented on conf_file.cpp:
```diff
@@ -16,268 +16,416 @@
 
-    memmove(str, s, end - s + 1);
+    memcpy(storage, val, val_len);
+    ((char *)storage)[val_len] = 0;
```
Minor, but `\0` is more clear about the intent of this `0`.

edersondisouza commented on conf_file.cpp:
```diff
@@ -16,268 +16,416 @@
+        c = _find_config(s, table[i].key, strlen(table[i].key));
+        if (!c) {
+            if (table[i].required) {
+                log_error("Required field '%s' not found in section '%.*s'", table[i].key,
```
Maybe inform last file on which this section was found? Not sure if showing the whole list if files in which this section appears is easy, but could be a plus, don't you think?

otaviobp replied:
We needed to keep a pointer to last file the section appeared. It is not hard, just some pointers added. A list to all files that included the section is a bit more complicated, but easy anyway.
If you think that this is a needed feature, I can added it. But I don't know if it is worth keeping it. Up to you.

edersondisouza replied:
Yeah, I've had an error here and had to grep to find which file caused the error. It's more of a nice to have... not sure if someone would have a long chain of overriding configs...

edersondisouza commented on main.cpp:
```diff
@@ -474,212 +460,174 @@ static const char *get_conf_dir()
 }
 
-static struct endpoint_config *search_endpoints(const char *name)
+static int parse_mavlink_dialect(const char *val, size_t val_len, void *storage, size_t storage_len)
```
World would be a better place if there were a easy way to generate an enum from string...

otaviobp replied:
I though about creating another struct to map this, but I didn't work on that. We could have this, but we would need a user_data pointer to be used in parser functions, with the mapping. Not sure if it is worth it.

lucasdemarchi replied:
> World would be a better place if there were a easy way to generate an enum from string...

well... this actually exists in several forms. P99 macros is one of them:
http://p99.gforge.inria.fr/p99-html/group__types_ga5d811e46f48e01d79d1f931817d9c0f1.html#ga5d811e46f48e01d79d1f931817d9c0f1

Other projects like systemd use code generation for similar thing.

lucasdemarchi replied:
and you will find solutions around using "modern c++XX" with some limitations.

edersondisouza replied:
See: 'several forms' and 'with some limitations' aren't exactly 'easy way' =D

lucasdemarchi replied:
eheheh... fair enough.

edersondisouza commented on conf_file.cpp:
```diff
@@ -16,268 +16,416 @@
+    if (fd < 0)
+        return -errno;
+
+    if (stat(filename, &fstat) < 0)
```
Hint: this trick can be handy https://github.com/01org/mavlink-router/blob/master/logendpoint.cpp#L101-L102

edersondisouza commented on conf_file.cpp:
```diff
@@ -16,268 +16,416 @@
-{
-    if (!_current_section->next)
-        return NULL;
+    end = (char *)memrchr(addr, ']', len);
```
Hmmm... `memrchr` allows things like `[really ] weird] option]`. Is it desirable?

otaviobp replied:
No. it is not. my bad.

otaviobp commented on conf_file.cpp:
```diff
@@ -16,268 +16,416 @@
+    ret = _parse_file((char *)addr, (size_t)fstat.st_size, filename);
+    if (ret < 0) {
+        log_error("Conf file parcially parsed. Configuration keys defined before the error were "
+                  "included");
```
I am just logging what has happened here and I am returning an error code.
I decided to print this because if you are parsing multiple files and the last one fails, some data from last file will be added anyway. I don't want people to assume that only data from previous files that were successfully parsed were added.

otaviobp commented on conf_file.cpp:
```diff
@@ -16,268 +16,416 @@
-        log_error("Invalid conf file %s", _filename);
-        return -EINVAL;
+    next:
+        len = len - (end - addr) - 1;
```
That's true. I need to subtract the original line len before trim. thanks

otaviobp commented:
Branched updated. All proposed fixes that I said that I would fix, or that were not commented were fixed.
I added some documentation on how to use ConfFile too.
And I removed the global_instance() static function because we won't need to use ConfFile as singleton in mavlink-router nor camera-streaming-daemon and it makes more sense to have it as a regular class.

edersondisouza commented on conf_file.cpp:
```diff
@@ -16,268 +16,412 @@
+    if (c) {
+        c->value = value;
+        c->value_len = value_len;
+    } else
```
Unbalanced braces should be avoided on mavlink-router - couldn't find a way to check for them on checkpatch..

edersondisouza commented on confs/b.conf:
```diff
@@ -0,0 +1,102 @@
+# mavlink-router configuration file is composed of sections,
```
A wild conf file appeared!!

edersondisouza commented on conf_file.cpp:
```diff
@@ -16,268 +16,412 @@
+                log_error(
+                    "On file %s: Line %d: Expected section before the definition of first option.",
+                    filename, line);
             }
```
Missing return or break here?

edersondisouza commented:
Another issue I've found: if I run mavlink-router without pointing to any config file, it just exits, no error, no nothing.

otaviobp commented:
Pushed new branch with all proposed fixes

edersondisouza commented on main.cpp:
```diff
@@ -474,212 +460,174 @@ static const char *get_conf_dir()
+    if (memcaseeq(val, val_len, "normal", sizeof("normal") - 1)) {
+        *eavesdropping = false;
+    } else if (memcaseeq(val, val_len, "eavesdropping", sizeof("eavesdropping") - 1)) {
+        *eavesdropping = false;
```
Guess here you wanted `true`

lucasdemarchi commented on conf_file.h:
```diff
@@ -17,24 +17,123 @@
 #pragma once
 
+#define DECLARE_PARSE_INT(_type) \
+    static int parse_##_type(const char *val, size_t val_len, void *storage, size_t storage_len)
```
this is only useful for "internal" usage by the class. You don't need this defined for who includes this header. As such, a better approach would be to move the define inside the class and after it's used you undef it.

lucasdemarchi commented on conf_file.h:
```diff
@@ -17,24 +17,123 @@
+     * @param ptr A void pointer to be used internally in the loop. Must be initialized with @c NULL
+     * for first execution.
+     */
+    int loop_sub_sections(const char *section_name, char separator, const char **sub_name,
```
separator will always be a space, no?

lucasdemarchi commented on conf_file.h:
```diff
@@ -17,24 +17,123 @@
+     */
+    int loop_sub_sections(const char *section_name, char separator, const char **sub_name,
+                          size_t *sub_len, const OptionsTable table[], size_t table_len, void *data,
+                          void **ptr);
```
I'm not sure about the approach on this function. I'd rather try to reuse extract_options() above, which would make most of these arguments go away.  The idea would be to have a `get_sections(const char *glob)` that accepts a glob and returns an iterator. You then loop on this iterator and call `extract_options()`. The "sub section" name is saved on this iterator.

I'd also avoid the "sub section" nomenclature. It's just a section IMO.

lucasdemarchi replied:
Maybe a change like this is very intrusive. If we agree on this, we can do it on top if you prefer.

lucasdemarchi commented on conf_file.h:
```diff
@@ -17,24 +17,123 @@
+/*
+ * Return the number of elements in a OptionTable array.
+ */
+#define OPTIONS_TABLE_SIZE(_table) (sizeof(_table) / sizeof(ConfFile::OptionsTable))
```
this is ARRAY_SIZE, no need to define it with another name.

lucasdemarchi commented on conf_file.h:
```diff
@@ -17,24 +17,123 @@
+#define OPTIONS_TABLE_SIZE(_table) (sizeof(_table) / sizeof(ConfFile::OptionsTable))
+
+/*
+ * Create the last two elements in OptionsTable, storage_offset and storage_len, using _field
```
"last two" is a good candidate for a forgotten comment when the table is changed

lucasdemarchi commented on conf_file.h:
```diff
@@ -17,24 +17,123 @@
+ * from _struct.
+ */
+#define OPTIONS_TABLE_STRUCT_FIELD(_struct, _field) \
+    offsetof(_struct, _field), sizeof(_struct::_field)
```
I think it's better to define OptionsTable with sub-struct containing offset and size. Then this would become:
```c++
#define OPTIONS_TABLE_STRUCT_FIELD(_struct, _field) \
     { offsetof(_struct, _field), sizeof(_struct::_field) }
```

This way it's easier to check that the number of fields initialized in the struct match the number of fields it should be.

otaviobp commented:
Branch updated with proposed fixes and new api for looping subsections

lucasdemarchi commented:
Great stuff!!

Thanks for all the rework. It looks like a really good API now.  I added 2 small commits on top with minor things/nitpicks. Applied.
